pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -113,6 +126,108 @@ func (p *provisioningAdmitter) Admit(request *admission.Request) (*admissionv1.A
+				Message: `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`,
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateDataDirectories will validate updates to the cluster object to ensure the data directories are not changed.
+// The only exception when a data directory is allowed to be changed is if cluster.Spec.AgentEnvVars has an env var with
+// for "CATTLE_AGENT_VAR_DIR", which Rancher will perform a one-time migration to set the
+// cluster.Spec.RKEConfig.DataDirectories.SystemAgent field for the cluster. validateAgentEnvVars will ensure
+// "CATTLE_AGENT_VAR_DIR" is not added, so this exception only applies to the one-time Rancher migration.
+func (p *provisioningAdmitter) validateDataDirectories(request *admission.Request, oldCluster, newCluster *v1.Cluster) *metav1.Status {
+	if request.Operation != admissionv1.Update {
+		return nil
+	}
+	if newCluster.Spec.RKEConfig == nil {
+		return nil
+	}
+	oldEnvVars := slices.DeleteFunc(slices.Clone(oldCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
+		return envVar.Name != "CATTLE_AGENT_VAR_DIR"
+	})
+	// will be true up until rancher performs one time migration to set SystemAgent data directory
+	if len(oldEnvVars) > 0 {
+		newEnvVars := slices.DeleteFunc(slices.Clone(newCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
+			return envVar.Name != "CATTLE_AGENT_VAR_DIR"
+		})
+		if reflect.DeepEqual(newEnvVars, oldEnvVars) && oldCluster.Spec.RKEConfig.DataDirectories.SystemAgent == newCluster.Spec.RKEConfig.DataDirectories.SystemAgent {
+			// rancher migration has not been performed yet
+			return nil
+		}
+		if newCluster.Spec.RKEConfig.DataDirectories.SystemAgent != oldEnvVars[len(oldEnvVars)-1].Value {
+			return &metav1.Status{
+				Status:  "Failure",
+				Message: `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`,
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			}
+		}
+	} else if oldCluster.Spec.RKEConfig.DataDirectories.SystemAgent != newCluster.Spec.RKEConfig.DataDirectories.SystemAgent {
+		return &metav1.Status{
+			Status:  "Failure",
+			Message: "SystemAgent data directory cannot be changed after cluster creation",
+			Reason:  metav1.StatusReasonBadRequest,
+			Code:    http.StatusBadRequest,
+		}
+	} else if oldCluster.Spec.RKEConfig.DataDirectories.Provisioning != newCluster.Spec.RKEConfig.DataDirectories.Provisioning {
+		return &metav1.Status{
+			Status:  "Failure",
+			Message: "Provisioning data directory cannot be changed after cluster creation",
+			Reason:  metav1.StatusReasonBadRequest,
+			Code:    http.StatusBadRequest,
+		}
+	} else if oldCluster.Spec.RKEConfig.DataDirectories.K8sDistro != newCluster.Spec.RKEConfig.DataDirectories.K8sDistro {
```

Comment:
Have you considered simplifying this into a single case? Something like:

```suggestion
	if reflect.DeepEqual(oldCluster.Spec.RKEConfig.DataDirectories, newCluster.Spec.RKEConfig.DataDirectories){
```

Reply:
Still have to leave separate but will become clear as to why when I push.